blockchain/blockchain.go

```diff
@@ -0,0 +1,198 @@
+package blockchain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/core/state"
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/utils"
+)
+
+type ErrIncompatibleBlockAndStateUpdate struct {
+	reason string
+}
+
+func (e ErrIncompatibleBlockAndStateUpdate) Error() string {
+	return fmt.Sprintf("incompatible block and state update: %v", e.reason)
+}
+
+type ErrIncompatibleBlock struct {
+	reason string
+}
+
+func (e ErrIncompatibleBlock) Error() string {
+	return fmt.Sprintf("incompatible block: %v", e.reason)
+}
+
+// BlockDbKey creates appends block number and hash to create a db key.
+// Todo: make private after Blockchain.BlockByNumber & BlockchainBlockByHash are implemented.
+type BlockDbKey struct {
+	Number uint64
+	Hash   *felt.Felt
+}
+
+func (k *BlockDbKey) MarshalBinary() ([]byte, error) {
+	var numB [8]byte
+	binary.BigEndian.PutUint64(numB[:], k.Number)
+	return db.Blocks.Key(numB[:], k.Hash.Marshal()), nil
+}
+
+func (k *BlockDbKey) UnmarshalBinary(data []byte) error {
+	if len(data) != 41 {
+		return errors.New("key should be 41 bytes long")
+	}
+
+	if data[0] != byte(db.Blocks) {
+		return errors.New("wrong prefix")
+	}
+
+	k.Number = binary.BigEndian.Uint64(data[1:9])
+	k.Hash = new(felt.Felt).SetBytes(data[9:41])
+	return nil
+}
+
+// Blockchain is responsible for keeping track of all things related to the StarkNet blockchain
+type Blockchain struct {
+	l *sync.RWMutex
+
+	head     *core.Block
+	network  utils.Network
+	database db.DB
+}
+
+func NewBlockchain(database db.DB, network utils.Network) (*Blockchain, error) {
+	chain := &Blockchain{
+		l:        &sync.RWMutex{},
+		database: database,
+		network:  network,
+	}
+	txn := chain.database.NewTransaction(false)
+	headBlockBin, err := txn.Get(db.HeadBlock.Key())
+	if errors.Is(err, db.ErrKeyNotFound) {
+		return chain, nil
+	}
+
+	headBlock := new(core.Block)
+	if err = gob.NewDecoder(bytes.NewReader(headBlockBin)).Decode(headBlock); err != nil {
+		return nil, err
+	}
+	chain.head = headBlock
+	return chain, nil
+}
+
+// Height returns the latest block height. If blockchain is empty nil is returned.
+func (b *Blockchain) Height() *uint64 {
```

Comment:
we are allocating a uint64 on the heap for no added benefit AFAICS. I think it is better to just return the value

Reply:
AFAIK escape analysis determines whether allocation will happen on the heap or stack. Since the variable will be referenced from outside the function I believe both the pointer and the value will be allocated on the heap. Also, the reason for this is to determine whether the blockchain is empty or not. If I were to return the value the caller will not be able to tell the difference between an empty blockchain or a blockchain with a current height of 0. 

Also, It is idiomatic in go to use a pointer to differentiate between zero and nil values.

Reply:
Due to the removal of the mutex. This piece of code has changed

Reply:
a `uint64` fits nicely in a CPU register. So it will be returned as a value rather than a reference. But if using `nil` makes sense, we can keep it as it is.